File: ports/go/internal/docproc/pipeline/probe.go
package pipeline

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/tensor"
)

// StageSink receives one named intermediate value per pipeline stage.
//
// Mirrors document_processing/pipeline/probe.py. Off by default: Pipeline holds a nil
// Sink and every emission site costs one nil check. Nothing may be computed for the
// sink's benefit — payloads are references to values the pipeline already produced —
// because instrumentation that changes the cost of the instrumented code is worse
// than none.
//
// Stage names are the fixed vocabulary in conformance/spec/stages.md. Additive only:
// goldens are keyed by them.
type StageSink interface {
	Emit(name string, payload any) error
}

// NullStageSink discards every emission.
//
// Mirrors probe.py's NullStageSink and exists for the same reason: it lets `recognize`
// share one code path with `probe` instead of maintaining a second walk of the pipeline
// that would be free to drift from it.
type NullStageSink struct{}

func (NullStageSink) Emit(string, any) error { return nil }

// ArrayPayload marks a payload that should be written as .npy rather than JSON.
//
// Go has no numpy, so the "is this an array?" question that Python answers with an
// isinstance check has to be explicit here. That is a deviation in mechanism, not in
// behaviour: the files produced are identical.
type ArrayPayload struct {
	Array *tensor.Array
}

// DirectoryStageSink writes each stage to <root>/<name>.npy or <name>.json, plus an
// ordered stages.json index — byte-compatible with the Python sink so the checker
// cannot tell which implementation produced a dump.
type DirectoryStageSink struct {
	root    string
	upto    string
	stopped bool
	index   []stageEntry
}

type stageEntry struct {
	Stage string `json:"stage"`
	File  string `json:"file"`
	Kind  string `json:"kind"`
	Dtype string `json:"dtype,omitempty"`
	Shape []int  `json:"shape,omitempty"`
}

// NewDirectoryStageSink creates the directory. An empty upto means "emit everything".
func NewDirectoryStageSink(root, upto string) (*DirectoryStageSink, error) {
	if err := os.MkdirAll(root, 0o755); err != nil {
		return nil, err
	}
	return &DirectoryStageSink{root: root, upto: upto}, nil
}

func (s *DirectoryStageSink) Emit(name string, payload any) error {
	if s.stopped {
		return nil
	}
	safe := filepath.Base(name) // stage names contain dots, never separators

	var entry stageEntry
	switch p := payload.(type) {
	case ArrayPayload:
		file := safe + ".npy"
		if err := tensor.Save(filepath.Join(s.root, file), p.Array); err != nil {
			return fmt.Errorf("probe: %s: %w", name, err)
		}
		entry = stageEntry{Stage: name, File: file, Kind: "npy",
			Dtype: dtypeName(p.Array.Dtype), Shape: p.Array.Shape}
	default:
		file := safe + ".json"
		blob, err := json.MarshalIndent(payload, "", "  ")
		if err != nil {
			return fmt.Errorf("probe: %s: %w", name, err)
		}
		if err := os.WriteFile(filepath.Join(s.root, file), append(blob, '\n'), 0o644); err != nil {
			return fmt.Errorf("probe: %s: %w", name, err)
		}
		entry = stageEntry{Stage: name, File: file, Kind: "json"}
	}
	s.index = append(s.index, entry)

	// --upto stops AFTER the named stage, so the stage asked about is included.
	if s.upto != "" && name == s.upto {
		s.stopped = true
	}
	return nil
}

// Count reports how many stages were written.
func (s *DirectoryStageSink) Count() int { return len(s.index) }

// Close writes the ordered index.
func (s *DirectoryStageSink) Close() error {
	payload := map[string]any{
		"upto":          nullableString(s.upto),
		"stopped_early": s.stopped,
		"stages":        s.index,
	}
	blob, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(s.root, "stages.json"), append(blob, '\n'), 0o644)
}

// nullableString keeps the index byte-compatible with Python's, which writes JSON
// null rather than "" for an absent --upto.
func nullableString(v string) any {
	if v == "" {
		return nil
	}
	return v
}

// dtypeName maps our dtype tags to the numpy spellings the index records
// ("uint8", not "|u1"), matching what np.save reports on the Python side.
func dtypeName(d tensor.DType) string {
	switch d {
	case tensor.Uint8:
		return "uint8"
	case tensor.Float32:
		return "float32"
	case tensor.Float64:
		return "float64"
	case tensor.Int64:
		return "int64"
	default:
		return string(d)
	}
}

File: document_processing/pipeline/probe.py
"""Optional per-stage instrumentation for cross-language conformance testing.

The library is being reimplemented in Go, .NET, Kotlin and C++. Comparing only
the final result of a port against this reference tells you *that* it diverged,
never *where* — and a whole-pipeline mismatch on a twelve-model pipeline is a
week of bisection. Comparing intermediate stages turns that into "first
divergence at ``fields.bbox``", which is an hour.

This module is the mechanism. ``Pipeline.probe`` is ``None`` in production, so
every emission site costs one attribute test and nothing else: no payload is
built, no copy is made, no branch is taken beyond the guard. Nothing here changes
what ``process_img`` computes or returns.

Two design constraints are deliberate and should survive refactoring:

* **Do not thread dump values through return types.** Making stage functions
  return extra data, or adding parameters they pass along, would alter the very
  code the ports are transliterating — and then the ports would be copying the
  instrumentation rather than the algorithm.
* **Emit references, not deep copies.** A sink that needs to keep a payload past
  the call is responsible for copying it. The pipeline reuses and rebinds arrays
  (``self.results`` is rebound on every ``process_img``), so a sink that stores
  a reference and reads it later gets whatever that array became — the same trap
  the service hit with ``PipelineResults``.

Stage names are a fixed, ordered vocabulary defined in
``conformance/spec/stages.md``. They may be added to but never renamed: the
golden files are keyed by them.

Usage::

    from document_processing.pipeline.probe import DirectoryStageSink

    with DirectoryStageSink('dump/') as sink:
        pipeline.probe = sink
        try:
            results = pipeline.process_img(path)
        finally:
            pipeline.probe = None
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Protocol, runtime_checkable

import numpy as np


@runtime_checkable
class StageSink(Protocol):
    """Receives one named intermediate value per pipeline stage.

    A single method on purpose. Anything richer (filtering, ordering, buffering)
    belongs in the implementation, not in the interface the pipeline depends on.
    """

    def emit(self, name: str, payload: Any) -> None:  # pragma: no cover - protocol
        ...


class NullStageSink:
    """A sink that discards everything.

    Exists so tests can exercise the emission sites without writing files.
    Production uses ``probe = None`` instead, which skips the calls entirely.
    """

    def emit(self, name: str, payload: Any) -> None:
        pass


class RecordingStageSink:
    """Keeps every payload in memory, in emission order.

    Payloads are stored **as passed**, without copying, matching the contract in
    this module's docstring. For arrays the pipeline may later rebind or mutate,
    pass ``copy=True``.
    """

    def __init__(self, copy: bool = False):
        self.copy = copy
        self.stages: list[tuple[str, Any]] = []

    def emit(self, name: str, payload: Any) -> None:
        if self.copy and isinstance(payload, np.ndarray):
            payload = payload.copy()
        self.stages.append((name, payload))

    def names(self) -> list[str]:
        return [n for n, _ in self.stages]

    def get(self, name: str) -> Any:
        for n, p in self.stages:
            if n == name:
                return p
        raise KeyError(name)


class DirectoryStageSink:
    """Writes each stage to ``<root>/<name>.npy`` or ``<name>.json``.

    ndarrays become ``.npy`` (self-describing: dtype, shape and order travel with
    the bytes, so a transposed payload fails loudly instead of comparing equal).
    Everything else becomes JSON, with numpy scalars and arrays coerced, because
    ``json.dumps`` cannot serialise them and a port would otherwise have nothing
    to compare against.

    An ordered index is written to ``stages.json`` so a consumer knows which
    stages ran — which is how ``--upto`` makes a partial port verifiable.
    """

    def __init__(self, root: str | Path, upto: str | None = None):
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)
        self.upto = upto
        self.index: list[dict[str, Any]] = []
        self._stopped = False

    def emit(self, name: str, payload: Any) -> None:
        if self._stopped:
            return

        safe = name.replace("/", "_")
        if isinstance(payload, np.ndarray):
            path = self.root / f"{safe}.npy"
            np.save(path, payload, allow_pickle=False)
            entry = {"stage": name, "file": path.name, "kind": "npy",
                     "dtype": str(payload.dtype), "shape": list(payload.shape)}
        else:
            path = self.root / f"{safe}.json"
            path.write_text(json.dumps(_jsonable(payload), indent=2, ensure_ascii=False),
                            encoding="utf-8")
            entry = {"stage": name, "file": path.name, "kind": "json"}
        self.index.append(entry)

        # `--upto` stops after the named stage rather than before, so the stage
        # the caller asked about is included.
        if self.upto is not None and name == self.upto:
            self._stopped = True

    def close(self) -> None:
        (self.root / "stages.json").write_text(
            json.dumps({"upto": self.upto, "stopped_early": self._stopped,
                        "stages": self.index}, indent=2, ensure_ascii=False),
            encoding="utf-8")

    def __enter__(self) -> "DirectoryStageSink":
        return self

    def __exit__(self, *exc) -> None:
        self.close()


def _jsonable(value: Any) -> Any:
    """Coerce numpy types so ``json.dumps`` succeeds without a ``default=`` hook.

    Deliberately explicit rather than relying on a fallback hook: an unexpected
    type should be visible here, not silently stringified into a golden file that
    a port can never reproduce.
    """
    if isinstance(value, np.ndarray):
        return value.tolist()
    if isinstance(value, np.generic):
        return value.item()
    if isinstance(value, dict):
        return {str(k): _jsonable(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_jsonable(v) for v in value]
    if isinstance(value, (str, int, float, bool)) or value is None:
        return value
    return repr(value)
